src/compiler/program.ts

```diff
@@ -958,6 +972,23 @@ namespace ts {
         return resolutions;
     }
 
+    export function getDefaultTypeDirectiveNames(options: CompilerOptions, rootFiles: string[], host: CompilerHost): string[] {
+        // Use explicit type list from tsconfig.json
+        if (options.types) {
+            return options.types;
+        }
+
+        // or load all types from the automatic type import fields
+        if (host && host.getDefaultTypeDirectiveNames) {
+            const commonRoot = computeCommonSourceDirectoryOfFilenames(rootFiles, host.getCurrentDirectory(), host.getCanonicalFileName);

```

Comment:
currently we are tearing off `this` which might be wrong in cases if custom version of `CompilerHost` uses it in its implementation of `getCanonicalFileName`. Probably instead we can make `computeCommonSourceDirectoryOfFilenames` to accept `{ getCanonicalFileName(s: string): string }` so we can just pass `host`/